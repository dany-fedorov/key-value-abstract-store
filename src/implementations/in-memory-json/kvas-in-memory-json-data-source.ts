import {
  KvasDataSource,
  KvasDataSourceInitializeOptions,
} from '@core/kvas-data-source';
import type {
  JsonComposite,
  JsonPrimitive,
  JsonValue,
} from '@in-memory-json/kvas-in-memory-json-types';
import type { KvasInMemoryJsonTypeParameters } from '@in-memory-json/kvas-in-memory-json-map';
import type { KvasInMemoryJsonMap } from '@in-memory-json/kvas-in-memory-json-map';
import { KvasInMemoryJsonMapOperations } from '@in-memory-json/kvas-in-memory-json-map-operations';
import { KvasDataSourceSyncContainer } from '@core/kvas-data-source-sync-container';
import { KvasDataSourceAsyncContainer } from '@core/kvas-data-source-async-container';

export type KvasInMemoryJsonDataSourceCreateOptions<P extends JsonPrimitive> = {
  operations?: KvasInMemoryJsonMapOperations<P>;
} & KvasDataSourceInitializeOptions<
  KvasInMemoryJsonTypeParameters<P>,
  KvasInMemoryJsonMap<P>,
  JsonValue<P>
>;

export class KvasInMemoryJsonDataSource<
  P extends JsonPrimitive,
> extends KvasDataSource<
  KvasInMemoryJsonTypeParameters<P>,
  KvasInMemoryJsonMap<P>,
  JsonComposite<P> | P
> {
  static createSyncDataStore<P extends JsonPrimitive>(
    options?: KvasInMemoryJsonDataSourceCreateOptions<P>,
  ) {
    const dataSource = new KvasDataSourceSyncContainer(
      new KvasInMemoryJsonDataSource<JsonPrimitive>(
        options?.operations ?? new KvasInMemoryJsonMapOperations<P>(),
      ),
    );
    dataSource.initialize(
      options && {
        ...(!options?.fromJSO ? {} : { fromJSO: options?.fromJSO }),
        ...(!options?.fromMap ? {} : { fromMap: options?.fromMap }),
      },
    );
    return dataSource;
  }

  static async createAsyncDataStore<P extends JsonPrimitive>(
    options?: KvasInMemoryJsonDataSourceCreateOptions<P>,
  ) {
    const dataSource = new KvasDataSourceAsyncContainer(
      new KvasInMemoryJsonDataSource<JsonPrimitive>(
        options?.operations ?? new KvasInMemoryJsonMapOperations<P>(),
      ),
    );
    await dataSource.initialize(
      options && {
        ...(!options?.fromJSO ? {} : { fromJSO: options?.fromJSO }),
        ...(!options?.fromMap ? {} : { fromMap: options?.fromMap }),
      },
    );
    return dataSource;
  }
}
