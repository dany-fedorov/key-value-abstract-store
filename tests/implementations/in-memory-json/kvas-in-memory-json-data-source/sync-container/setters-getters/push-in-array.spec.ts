import { KvasInMemoryJsonDataSource } from '@in-memory-json/kvas-in-memory-json-data-source';

it('should push to object selecting key automatically', () => {
  const ds = KvasInMemoryJsonDataSource.createSyncDataStore();
  const path = ['b', 'c'];
  const value = 'the-value';
  const { key: key1 } = ds.push(path, value);
  const { key: key2 } = ds.push(path, value + 2);
  const { key: key3 } = ds.push(path, value + 3);
  expect({ key1, key2, key3 }).toMatchSnapshot();
  expect(ds.getJSO(path)).toMatchSnapshot();
  expect(ds).toMatchSnapshot();
});
