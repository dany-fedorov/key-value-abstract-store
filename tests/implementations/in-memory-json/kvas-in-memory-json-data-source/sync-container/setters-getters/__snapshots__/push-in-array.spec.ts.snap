// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should push to object selecting key automatically 1`] = `
Object {
  "key1": 0,
  "key2": 1,
  "key3": 2,
}
`;

exports[`should push to object selecting key automatically 2`] = `
Object {
  "prop": Object {
    "path": Array [
      "b",
      "c",
    ],
    "type": "map",
    "value": Array [
      "the-value",
      "the-value2",
      "the-value3",
    ],
  },
}
`;

exports[`should push to object selecting key automatically 3`] = `
KvasDataSourceSyncContainer {
  "dataSource": KvasInMemoryJsonDataSource {
    "operations": KvasInMemoryJsonMapOperations {},
    "rootMap": KvasInMemoryJsonMap {
      "host": Object {
        "b": Object {
          "c": Array [
            "the-value",
            "the-value2",
            "the-value3",
          ],
        },
      },
      "icfg": Object {
        "getValueType": [Function],
        "options": Object {
          "ignoreIncompatibleHostAndKey": false,
        },
      },
      "mapType": "object",
    },
  },
}
`;
